win32/tests/file_win32_ut/file_win32_ut.c

```diff
@@ -0,0 +1,952 @@
+    STRICT_EXPECTED_CALL(mock_CreateThreadpoolIo(handle, IGNORED_ARG, NULL, IGNORED_ARG))
+        .CaptureArgumentValue_pcbe(&captured_pcbe)
+        .CaptureArgumentValue_pfnio(captured_callback)
+        .SetReturn(ptp_io);
+
+    ///act
+    FILE_HANDLE file_handle = file_create(execution_engine, filename, NULL, NULL);
+
+    ///assert
+    ASSERT_IS_NOT_NULL(file_handle);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    umock_c_reset_all_calls();
+    return file_handle;
+}
+
+static void get_callback(FILE_CB user_callback, void* user_context, uint32_t size, PTP_WIN32_IO_CALLBACK* captured_callback, LPOVERLAPPED* captured_ov)
+{
+    ///arrange
+    FILE_HANDLE file_handle = get_file_handle("test_file.txt", captured_callback);
+    unsigned char buf[10];
+    HANDLE h_event = (HANDLE)44;
+    uint64_t position = 5;
+    PTP_IO captured_ptp_io;
+
+    STRICT_EXPECTED_CALL(malloc(IGNORED_ARG));
+    STRICT_EXPECTED_CALL(mock_CreateEvent(IGNORED_ARG, IGNORED_ARG, IGNORED_ARG, IGNORED_ARG))
+        .SetReturn(h_event);
+    STRICT_EXPECTED_CALL(mock_StartThreadpoolIo(IGNORED_ARG))
+        .CaptureArgumentValue_pio(&captured_ptp_io);
+
+    STRICT_EXPECTED_CALL(mock_WriteFile(IGNORED_ARG, buf, size, NULL, IGNORED_ARG))
+        .CaptureArgumentValue_lpOverlapped(captured_ov)
+        .SetReturn(FALSE);
+    STRICT_EXPECTED_CALL(mock_GetLastError())
+        .SetReturn(ERROR_IO_PENDING);
+
+    ///act
+    FILE_WRITE_ASYNC_RESULT result = file_write_async(file_handle, buf, size, position, user_callback, user_context);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_OK, result);
+    
+    //cleanup
+    file_destroy(file_handle);
+    umock_c_reset_all_calls();
+}
+
+BEGIN_TEST_SUITE(file_win32_unittests)
+
+TEST_SUITE_INITIALIZE(suite_init)
+{
+    g_testByTest = TEST_MUTEX_CREATE();
+    ASSERT_IS_NOT_NULL(g_testByTest);
+    ASSERT_ARE_EQUAL(int, 0, umock_c_init(on_umock_c_error));
+    ASSERT_ARE_EQUAL(int, 0, umocktypes_windows_register_types());
+    ASSERT_ARE_EQUAL(int, 0, umocktypes_bool_register_types());
+
+    REGISTER_GLOBAL_MOCK_HOOK(malloc, real_malloc);
+    REGISTER_GLOBAL_MOCK_HOOK(free, real_free);
+
+    REGISTER_UMOCK_ALIAS_TYPE(PTP_IO, void*);
+    REGISTER_UMOCK_ALIAS_TYPE(PTP_CALLBACK_ENVIRON, void*);
+    REGISTER_UMOCK_ALIAS_TYPE(EXECUTION_ENGINE_HANDLE, void*);
+    REGISTER_UMOCK_ALIAS_TYPE(PTP_POOL, void*);
+    REGISTER_UMOCK_ALIAS_TYPE(PTP_CLEANUP_GROUP, void*);
+
+    REGISTER_TYPE(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_RESULT);
+    REGISTER_TYPE(FILE_READ_ASYNC_RESULT, FILE_READ_ASYNC_RESULT);
+
+    REGISTER_GLOBAL_MOCK_FAIL_RETURN(malloc, NULL);
+    REGISTER_GLOBAL_MOCK_FAIL_RETURN(mock_CreateEvent, NULL);
+
+    REGISTER_GLOBAL_MOCK_FAIL_RETURN(mock_CreateThreadpoolCleanupGroup, NULL);

```

Comment:
>REGISTER_GLOBAL_MOCK_FAIL_RETURN [](start = 4, length = 32)

most of these can be REGISTER_GLOBAL_MOCK_RETURNS as I mentioned and they can set the fixed value to return here once.